File: src/Systems/Objective-C/example3.m
#import <Foundation/Foundation.h>

// Abstract Product
@protocol Report <NSObject>
- (void)generate;
@end

// Concrete Product
@interface PDFReport : NSObject <Report>
@end

@implementation PDFReport
- (void)generate {
    NSLog(@"Generating PDF report");
}
@end

@interface HTMLReport : NSObject <Report>
@end

@implementation HTMLReport
- (void)generate {
    NSLog(@"Generating HTML report");
}
@end

// Abstract Factory
@protocol ReportFactory <NSObject>
- (id<Report>)createReport;
@end

// Concrete Factory
@interface PDFReportFactory : NSObject <ReportFactory>
@end

@implementation PDFReportFactory
- (id<Report>)createReport {
    return [PDFReport new];
}
@end

@interface HTMLReportFactory : NSObject <ReportFactory>
@end

@implementation HTMLReportFactory
- (id<Report>)createReport {
    return [HTMLReport new];
}
@end

void useFactory(id<ReportFactory> factory) {
    id<Report> report = [factory createReport];
    [report generate];
}

int main(int argc, const char * argv[]) {
    @autoreleasepool {
        useFactory([PDFReportFactory new]);
        useFactory([HTMLReportFactory new]);
    }
    return 0;
}

File: src/Enterprise/VisualBasic/Example3.vb
Imports System

' Abstract Product
Public Interface Report
    Sub Generate()
End Interface

' Concrete Product
Public Class PDFReport
    Implements Report
    Public Sub Generate() Implements Report.Generate
        Console.WriteLine("Generating PDF report")
    End Sub
End Class

Public Class HTMLReport
    Implements Report
    Public Sub Generate() Implements Report.Generate
        Console.WriteLine("Generating HTML report")
    End Sub
End Class

' Abstract Factory
Public Interface ReportFactory
    Function CreateReport() As Report
End Interface

' Concrete Factory
Public Class PDFReportFactory
    Implements ReportFactory
    Public Function CreateReport() As Report Implements ReportFactory.CreateReport
        Return New PDFReport()
    End Function
End Class

Public Class HTMLReportFactory
    Implements ReportFactory
    Public Function CreateReport() As Report Implements ReportFactory.CreateReport
        Return New HTMLReport()
    End Function
End Class

Module Example3
    Sub UseFactory(factory As ReportFactory)
        Dim report As Report = factory.CreateReport()
        report.Generate()
    End Sub

    Sub Main()
        UseFactory(New PDFReportFactory())
        UseFactory(New HTMLReportFactory())
    End Sub
End Module
